src/dialects/oracle/query-generator.js

```diff
@@ -0,0 +1,1443 @@
+      throw new Error(`Invalid json statement: ${stmt}`);
+    }
+
+    // return true if the statement has valid json function
+    return hasJsonFunction;
+  }
+
+  jsonPathExtractionQuery(column, path) {
+    let paths = _.toPath(path);
+    const quotedColumn = this.isIdentifierQuoted(column) ? column : this.quoteIdentifier(column);
+
+    paths = paths.map(subPath => {
+      return /\D/.test(subPath) ? Utils.addTicks(subPath, '"') : subPath;
+    });
+
+    const pathStr = this.escape(['$'].concat(paths).join('.').replace(/\.(\d+)(?:(?=\.)|$)/g, (__, digit) => `[${digit}]`));
+
+    return `json_value(${quotedColumn},${pathStr})`;
+  }
+
+  addLimitAndOffset(options, model) {
+    let fragment = '';
+    const offset = options.offset || 0,
+      isSubQuery = options.hasIncludeWhere || options.hasIncludeRequired || options.hasMultiAssociation;
+
+    let orders = {};
+    if (options.order) {
+      orders = this.getQueryOrders(options, model, isSubQuery);
+    }
+
+    if (options.limit || options.offset) {
+      if (!(options.order && options.group) && (!options.order || options.include && !orders.subQueryOrder.length)) {
+        fragment += options.order && !isSubQuery ? ', ' : ' ORDER BY ';
+        fragment += `${this.quoteTable(options.tableAs || model.name) }.${this.quoteIdentifier(model.primaryKeyField)}`;
+      }
+
+      if (options.offset || options.limit) {
+        fragment += ` OFFSET ${this.escape(offset)} ROWS`;
+      }
+
+      if (options.limit) {
+        fragment += ` FETCH NEXT ${this.escape(options.limit)} ROWS ONLY`;
+      }
+    }
+
+    return fragment;
+  }
+
+  booleanValue(value) {
+    return value ? 1 : 0;
+  }
+
+  quoteIdentifier(identifier, force) {
+    const optForceQuote = force || false;
+    const optQuoteIdentifiers = this.options.quoteIdentifiers !== false;
+    const rawIdentifier = Utils.removeTicks(identifier, '"');
+    const regExp = (/^(([\w][\w\d_]*))$/g);
+
+    if (
+      optForceQuote !== true &&
+      optQuoteIdentifiers === false &&
+      regExp.test(rawIdentifier) === true &&
+      !ORACLE_RESERVED_WORDS.includes(rawIdentifier.toUpperCase())
+    ) {
+      // In Oracle, if tables, attributes or alias are created double-quoted,
+      // they are always case sensitive. If they contain any lowercase
+      // characters, they must always be double-quoted otherwise it
+      // would get uppercased by the DB.
+      // Here, we strip quotes if we don't want case sensitivity.
+      return rawIdentifier;
+    }
+    return Utils.addTicks(rawIdentifier, '"');
+  }
+
+  /**
+ * It causes bindbyPosition like :1, :2, :3
+ * We pass the val parameter so that the outBind indexes
+ * starts after the inBind indexes end
+   *
+ * @param {Array} bind
+ * @param {number} posOffset
+ */
+  bindParam(bind, posOffset = 0) {
+    return value => {
+      bind.push(value);
+      return `:${bind.length + posOffset}`;
+    };
+  }
+}
+
+// private methods
+function wrapSingleQuote(identifier) {
```

Comment:
I am only now realizing what `wrapSingleQuote` does. It removes all `'` characters.

```
wrapSingleQuote(`a'b`)
> 'ab'
```

I think 99% of the time you'' want to use `this.escape()` instead of `wrapSingleQuote`.

This is an issue I'll have to look into for mssql/db2.
The same problem is present with `Utils.addTicks`. It mangles values instead of escaping them

Reply:
Yes, this.escape works well for us. Removed the wrapSingleQuote function.